spark/src/main/java/org/apache/iceberg/actions/RemoveOrphanFilesAction.java

```diff
@@ -254,4 +270,53 @@ private static void listDirRecursively(
       return files.iterator();
     };
   }
+
+  protected static List<String> findOrphanFiles(
+      Dataset<Row> validFileDF,
+      Dataset<Row> actualFileDF) {
+    Column nameEqual = filename.apply(actualFileDF.col(RELATIVE_FILE_PATH))
+        .equalTo(filename.apply(validFileDF.col(RELATIVE_FILE_PATH)));
+
+    Column pathContains = actualFileDF.col(RELATIVE_FILE_PATH)
+        .contains(validFileDF.col(RELATIVE_FILE_PATH));
+
+    Column joinCond = nameEqual.and(pathContains);
+    Column decodeFilepath = decode.apply(actualFileDF.col(FILE_PATH));
+    return actualFileDF.join(validFileDF, joinCond, "leftanti").select(decodeFilepath)
+        .as(Encoders.STRING())
+        .collectAsList();
+  }
+
+  /**
+   * From
+   * <pre>{@code
+   * Dataset<Row<file_path_with_scheme_authority>>
+   *    will be transformed to
+   *    Dataset<Row<file_path_no_scheme_authority, file_path_with_scheme_authority>>
+   *  }</pre>
+   *
+   * This is required to compare the valid and all files to find the orphan files.
+   * Based on the result data set, only path will be compared while comparing valid and all files path.
+   * As in the case of hadoop, s3, there could be different authority names to access same path, which can give us files
+   * which are part of metadata and not orphan.
+   *
+   * @param filePathWithSchemeAndAuthority : complete file path, can include scheme, authority and path.
+   * @return : {@code file_path_no_scheme_authority, file_path}
+   */
+  protected static Dataset<Row> addRelativePathColumn(Dataset<Row> filePathWithSchemeAndAuthority) {
+    return filePathWithSchemeAndAuthority.withColumn(URI_DETAIL,
+        fileWithRelativePath.apply(
```

Comment:
This is where the UDF naming looks confusing to me since we call apply on 2 objects which are very different and they have similar names, one is data one is code.